### PR TITLE
LUCENE-9971: Inconsistent SSDVFF and Taxonomy facet behavior in case of unseen dimension

### PR DESCRIPTION
**Pull request by sqshq**
# Description

Unlike other `org.apache.lucene.facet.Facet` interface implementations, `SortedSetDocValuesFacetCounts` and `ConcurrentSortedSetDocValuesFacetCounts` throw an exception in case of unseen dimension in the `getTopChildren()` method.

It is inconsistent with [Facet interface documentation](https://github.com/apache/lucene/blob/main/lucene/facet/src/java/org/apache/lucene/facet/Facets.java#L33-L34) and not pleasant to deal with: in our case, the requested dimensions can be manually specified by the user, so we should always catch that and check the exception message to be sure of the exact cause (exception as a control flow anti-pattern).

Hopefully, it is not a big backward compatibility concern. [Current implementation may return null](https://github.com/sqshq/lucene/blob/main/lucene/facet/src/java/org/apache/lucene/facet/sortedset/SortedSetDocValuesFacetCounts.java#L140) before these changes, so users should be prepared for this.

UPD:
During the PR review, it was also decided to make similar changes for Taxonomy implementation.

# Solution

For SSDVFF, return `null` instead of throwing exceptions in case of unseen dimension. For Taxonomy, check that the dimension has ever been customized and throw unseen dimension IAE only in this case. 
The changes also include minor refactoring/cleanup.

# Tests

- Unit tests in `org.apache.lucene.facet` package

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/lucene/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Lucene maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.

### CODE REVIEW

**Review by gsmiller — COMMENTED** [review]
Interesting change! I think I'm a little confused about the consistency implications here. From a quick glance at the taxonomy-based code, I think you can end up with IllegalArgumentExceptions there as well if requesting counts for a dim that wasn't indexed. Specifically, if you're using the non-default index field and ask for a dim that was never indexed (see TaxonomyFacets#verifyDim). It looks like you might get away with requesting an invalid dim there if using the default index field, which seems more accidental than intentional (maybe that should be addressed?). Then again, I didn't write any code to confirm this behavior, so maybe I'm totally wrong.

Anyway, back to this specific change, I think it's reasonable to return null (it's much like a Java Map). Could you look into the taxonomy stuff though as part of this and see what happens if you're using the non-default field and try requesting a non-indexed dim? If consistency is a part of the goal here, it'd be nice if that could be consistent.

**Comment by sqshq** [comment]
Hi Greg,
You are right, [`TaxonomyFacets` fail with `IllegalArgumentException`](https://github.com/sqshq/lucene/blob/7f8b7ffbcad2265b047a5e2195f76cc924028063/lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java#L121) in the case when we are trying to search for a field that has a different `indexFieldName`. Although here we are not even checking that the `dimension` is present for the configured `indexFieldName` - this error is only related to the field mismatch.

That behavior actually looks more or less consistent with SSDVFF implementation. There, if we try to create a `SortedSetDocValuesReaderState` for a field that was never indexed with facets, [we fail with `IllegalArgumentException`](https://github.com/sqshq/lucene/blob/7f8b7ffbcad2265b047a5e2195f76cc924028063/lucene/facet/src/java/org/apache/lucene/facet/sortedset/DefaultSortedSetDocValuesReaderState.java#L73).

Although those `IllegalArgumentExceptions` are thrown in slightly different situations, I think in both cases they are pointing to a field name misconfiguration (or a bug in the code), rather than "user requested unseen dimension" scenario. As far as I understand in most of the cases the facet field name is a single/constant set of values, so most likely we indeed want to make sure that user does not ignore it. Also, this behavior can't be easily changed for SSDVFF, since the field check happens in `SortedSetDocValuesReaderState` constructor - we can't return null there. 

@gsmiller let me know what do you think!

**Comment by gsmiller** [comment]
Thanks for diving into this @sqshq! Yeah, I think I agree with all this. I suppose the subtle difference here is that in taxonomy-based facet counting, `FacetsConfig` would never have been told about a custom index field (because presumably it's not a valid dimension). You could get in a weird scenario though where you're using some custom index field for all your taxonomy faceting (and have never indexed anything in `$facets`). In other words, you've specified one or more custom index fields to hold the doc values via `FacetsConfig` and your index doesn't contain the `$facets` field at all. In this (somewhat unusual) case, if calling code were to ask for a dimension that was never indexed (same scenario you're describing in SSDVFC), you'd still get an `IllegalArgumentException` in `TaxonomyFacets#verifyDim` (because it would try to look up the configured index field for the dimension, wouldn't find it, fall back to `$facets` and then also not find that). That said, I don't think fixing that as well necessarily needs to be part of this change. It would be nice though if that could also consistently return `null`.

Well, +1 for making the change you proposed, particularly since it's consistent with the Javadoc as you point out as well.

**Review by gsmiller — APPROVED** [review]

**Comment by sqshq** [comment]
@gsmiller I got your point, and certainly don't mind making the Taxonomy part consistent for this edge case as well. 
Please take a look: https://github.com/apache/lucene/pull/149/commits/fd01e65c45b16cf94098aa2b199bd34494d192be

File: lucene/facet/src/test/org/apache/lucene/facet/taxonomy/TestTaxonomyFacetAssociations.java [hunk 1]
```diff
@@ -168,17 +169,8 @@ public void testWrongIndexFieldName() throws Exception {
...
-        () -> {
-          facets.getSpecificValue("float");
-        });
-
-    expectThrows(
-        IllegalArgumentException.class,
-        () -> {
-          facets.getTopChildren(10, "float");
-        });
+    Assert.assertEquals(-1, facets.getSpecificValue("float"));
```
Comment (gsmiller) [inline lucene/facet/src/test/org/apache/lucene/facet/taxonomy/TestTaxonomyFacetAssociations.java:172]: You should be able to leave off `Assert.` in these checks (and remove the import as well). These test cases actually sub-class `Assert` itself, so the import and `Assert.` bit isn't necessary

**Comment by gsmiller** [comment]
Thanks for taking this on @sqshq!

File: lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java [hunk 2]
```diff
@@ -110,17 +109,10 @@ public boolean siblingsLoaded() {
     return children != null;
   }
 
-  /**
-   * Throws {@code IllegalArgumentException} if the dimension is not recognized. Otherwise, returns
-   * the {@link DimConfig} for this dimension.
-   */
-  protected FacetsConfig.DimConfig verifyDim(String dim) {
+  /** Returns false if the dimension was not indexed into {@link #indexFieldName} field. */
+  protected boolean isDimIndexed(String dim) {
```
[the remark this answers is not in the document]
Comment (sqshq) [reply lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java]: Yes, I like the idea and it seems that would provide a good balance between prior strict assertion and "always return `null`" solution. What do you think about this implementation? https://github.com/apache/lucene/commit/38e54f73f00b518a617ab6f6231639adcd515aa2

File: lucene/facet/src/java/org/apache/lucene/facet/taxonomy/IntTaxonomyFacets.java [hunk 3]
```diff
@@ -133,7 +133,10 @@ private int rollup(int ord) throws IOException {
 
   @Override
   public Number getSpecificValue(String dim, String... path) throws IOException {
-    DimConfig dimConfig = verifyDim(dim);
+    if (!isDimIndexed(dim)) {
```
[the remark this answers is not in the document]
Comment (sqshq) [reply lucene/facet/src/java/org/apache/lucene/facet/taxonomy/IntTaxonomyFacets.java]: Thanks for pointing that out, I fixed that! Btw might be a good idea to add this rule to Lucene's check style settings, so `./gradlew check` can spot this automatically?

File: lucene/facet/src/test/org/apache/lucene/facet/taxonomy/TestTaxonomyFacetAssociations.java (hunk 1 above)
Comment (sqshq) [reply lucene/facet/src/test/org/apache/lucene/facet/taxonomy/TestTaxonomyFacetAssociations.java]: :+1: these asserts were removed with the recent changes, but I checked that there is no more `Assert.` anywhere in my code.

File: lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java [hunk 4]
```diff
@@ -157,6 +157,11 @@ public DimConfig getDimConfig(String dimName) {
     return dimConfig;
   }
 
+  /** Returns false if the dimension was never configured. */
```
Comment (gsmiller) [inline lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java:160]: Maybe add a small note in the javadoc that this doesn't necessarily mean the dimension is "invalid", but just that it will use all "default" configuration? I could see a reader of this javadoc interpreting it to mean that the dimension is "invalid".

File: lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java [hunk 5]
```diff
@@ -16,15 +16,17 @@
  */
 package org.apache.lucene.facet.taxonomy;
 
+import static org.apache.lucene.facet.FacetsConfig.*;
```
Comment (gsmiller) [inline lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java:19]: From what I've seen, Lucene avoid static imports (probably another instance of automated checks being useful).

File: lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java [hunk 6]
```diff
@@ -111,14 +113,25 @@ public boolean siblingsLoaded() {
   }
 
   /**
-   * Throws {@code IllegalArgumentException} if the dimension is not recognized. Otherwise, returns
-   * the {@link DimConfig} for this dimension.
+   * Verifies and returns {@link DimConfig} the given dimension name.
```
Comment (gsmiller) [inline lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java:116]: I think you meant to say "returns {@link DimConfig} for the given dimension name." (left out the word "for")?

**Review by gsmiller — COMMENTED** [review]
This looks great! I left a couple very small comments. Thanks again!

File: lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java (hunk 4 above)
Comment (sqshq) [reply lucene/facet/src/java/org/apache/lucene/facet/FacetsConfig.java]: Sounds good, added these details to the javadoc: https://github.com/apache/lucene/pull/149/commits/074939baa1fca6be46eb33542c7aa76cf4cddafa. We can also rename the method to `isDimCustomized`, if you think it looks less ambiguous. Let me know!

File: lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java (hunk 5 above)
Comment (sqshq) [reply lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java]: Sure, good to know! https://github.com/apache/lucene/pull/149/commits/074939baa1fca6be46eb33542c7aa76cf4cddafa

File: lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java (hunk 6 above)
Comment (sqshq) [reply lucene/facet/src/java/org/apache/lucene/facet/taxonomy/TaxonomyFacets.java]: Fixed that, thanks! https://github.com/apache/lucene/pull/149/commits/074939baa1fca6be46eb33542c7aa76cf4cddafa

**Comment by gsmiller** [comment]
Looks great @sqshq! Thanks for taking this up.
